### DISCUSSION TITLE
Some experiences and feedback

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#410] [General]
We've been doing an XR promotion tour, mostly at universities, over the past months, and one of the applications I showed (on our Varjo XR-3) was a simple 3D model viewer with a set of photogrammetry models the user could choose from. I think I've had around 100-150 folks with a research and/or (higher) education background try this demo in that time period. Here's some observations that might be interesting, plus some unsollicited feedback ;-)

- StereoKit was really stable! I did not have any crashes at all, as far as I can remember. Also, the very quick startup time (1-2 seconds) made it really easy (and preferable) to run, compared to the Unity-based Varjo demos that usually took 10-15 seconds to get running.
  
- Performance of the SK app was excellent, 90 fps (on an RTX3090 ;-)) in almost all cases, even when using large (3M+ poly) textured models. All models were loaded locally, and only for the really large ones was there a noticeable short period of application lockup during loading (I think this is the GPU texture upload, not being async yet?).

- Depth perception, especially when it came to interaction with the main menu (which held a bunch of large buttons), was an issue. Buttons were further away than most people thought ("take another step forward"), and a lot of them struggled with clicking the buttons. The circle marker shown on a button just before clicking is not really intuitive and probably too subtle. Maybe highlight the button in a different color? I.e. if the finger moves close the button changes from grey to red(dish) depending on distance, to indicate the button is about to be pressed.
  
- I disabled the far interaction of the menus, as it's too easy to accidentally activate/grab the menu otherwise. As the goal of the app was to allow lay-people to be able to use it without much instruction I wanted it to be as robust as possible and confuse folks with more advanced interaction that might give them a worse experience. In this respect the hands-first approach of SK is really powerful, because as soon folks see that their hands are tracked they understand the concept. 
  
- I use a set of option buttons attached to the left hand (when palm up) to be able to switch viewpoints. For some reason a lot of people kept trying to press the option that was currently active, instead of one of the inactive ones. This might be due to the very similar grey color of active and inactive buttons. Also, the amount of depression of the active button is pretty subtle. Perhaps they thought pressing a button did not do anything and so kept trying over and over. Edit: it also takes a bit of experience to learn how to effectively interact with UI elements using tracked hands, of course, but there might still be tweaks to the UI elements that can help there.
  
- Some of the points above were influenced by issues outside of SK's control, specifically the ["fat finger"](https://github.com/StereoKit/StereoKit/discussions/337) issue (due to an incorrect finger radius reported by the OpenXR layer) and hand tracking issues ([incorrect values reported](https://github.com/StereoKit/StereoKit/discussions/363) when a hand leaves the tracking volume). The latter actually caused me to disable all model grabbing interaction, as it simply wasn't usable. I can come up with specific hacks in SK code for both of these, but they seem like too specific to include in SK itself? It's interesting that these issues are not present when used through Unity or Unreal (which access the UltraLeap tracker through a non-OpenXR API), so these currently are OpenXR-specific issues, sadly. So far, these haven't been fixed at the OpenXR UltraLeap layer yet, which makes me curious if we'll see the same issues on the upcoming Lynx R1 (when used through OpenXR). 

- SK currently does not seem to take advantage of the full resolution of the XR-3, due to the high-res inset in the middle of the display needing a separate render pass. At least, I think that's currently not handled by SK, based on what the Varjo tools report when the application is running. This was not really a problem, as the lower-res full display is impressive already, which has quite a bit more pixels than, say, a Quest 2. But it would take an XR-3 specific extension to the SK code to handle.
 
- When folks wear an XR headset they suddenly seem to forget which hand is which, really weird. After explaining the viewpoint control menu was on their left hand I've had to point out to many people that is was on their "other" left hand.

- Development with SK is a breeze! I've dabbled in Unity a bit over the years, and tried Unreal at some point, but it takes quite a bit of time investment to get into and keep up with developments of those engines. In contrast, SK has a very low threshold to get started (and keep going), if you're already a bit familiar with 3D graphics and C#. I also love the immediate-mode way of doing things. Especially for smaller projects this is so much better and more intuitive than having to adhere to the straight-jacket that is "someone else's event loop".
  
- In terms of wish-list for SK I'd say somewhat more realistic rendering (particularly shadows, or perhaps a built-in AO shader) would be a good addition. Some of the other things I would be interested in are already planned, i.e. more UI theming support, and being able to [force controller usage](https://github.com/StereoKit/StereoKit/issues/364) and [more controller support](https://github.com/StereoKit/StereoKit/issues/150), and being able to switch [pass-through/opaque mode](https://github.com/StereoKit/StereoKit/issues/342) at run-time.

**Comment by maluoi** [comment]
Hey thank you _so much_ for taking the time to write this up! I'm definitely going to bookmark this and re-read it a few times while re-working the UI :D I haven't had as much of an opportunity to do live demos with SK since the pandemic, so this is definitely worth its weight in gold to me! Fantastic sample size too, that must have been a lot of demoing!

I don't think I have the time to write as much as I'd like in response to these, so here's a few select things :)

The XR-3 resolution thing I expect is the `XR_VARJO_quad_views` extension? I'm really glad SK has been working without it so far! I'll see if I can take a peek at that in the near future, and see how hard it is to add in :)

The stutter you were seeing with loading models was likely the mesh upload! This part is actually still synchronous, but is generally _way_ faster than textures, so not as noticeable. However for photogrammetry, there tends to be _a lot_ of vertices compared to normal meshes, so the upload becomes more noticeable there! This'll definitely be a soon thing :D 

And with respect to your wish-list, I'm very interested in adding lighting estimation to StereoKit! I've got plans for a render graph that'll enable a lot more interesting rendering tricks like this :) Unfortunately these are probably a few updates away, my current focus is on a UI/interaction revamp and WebXR support. So I hope to be addressing a lot of your UX concerns here in the near-term, and I'm definitely looking forward to more graphics code further in the future!
